src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPath.cs

```diff
@@ -82,6 +82,19 @@ public int Count
             get { return this.segments.Count; }
         }
 
+        /// <summary>
+        /// Get the segment in the path at the specified index.
+        /// </summary>
+        /// <param name="i">The index.</param>
+        /// <returns>The segment corresponding to the index.</returns>
+        public ODataPathSegment this[int i]

```

Comment:
I'd like to understand what we are actually trying to get out of this. At this point, it seems like `ODataPath` started out not trying to leak implementation details, but now we might as well have it implement `IReadOnlyList<ODataPathSegment>`

Reply:
I just needed a way to access the segments other than `First` and `Last` without enumerating through. Implementing `IReadOnlyList<ODataPathSegment>` would be a good idea - let me do that and we'll see how it looks.